unity/Assets/QuestNav/Native/NTCore/StringArrayPublisher.cs

```diff
@@ -0,0 +1,82 @@
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace QuestNav.Native.NTCore
+{
+    public unsafe class StringArrayPublisher
+    {
+        private readonly uint handle;
+        private string[] currentValue;
+
+        internal StringArrayPublisher(uint handle)
+        {
+            this.handle = handle;
+            this.currentValue = Array.Empty<string>();
+        }
+
+        private bool HasValueChanged(string[] newValue)
+        {
+            if (currentValue.Length != newValue.Length)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < newValue.Length; i++)
+            {
+                if (!currentValue[i].Equals(newValue[i]))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public bool Set(string[] values)
+        {
+            values ??= Array.Empty<string>();
+            // Avoid unnecessary marshaling & allocation for unchanged values
+            if (!HasValueChanged(values))
+            {
+                return false;
+            }
+
+            bool result;
+            WpiString[] wpiStrings = new WpiString[values.Length];
+
+            try
+            {
+                for (int i = 0; i < values.Length; i++)
```

Comment:
We loop through the managed strings, alloc some unmanaged memory, then copy the contents of the managed string over. In the `finally` clause we always free any `str`s that aren't `nullptr`. This whole process can be slow so we cache the last value, skip the expensive call if we `Set` the StringArray to the same values.